### PR TITLE
feature[next]: Use global type information in CollapseTuple pass

### PR DESCRIPTION
**Pull request by tehrengruber**
Previously the `CollapseTuple` Pass ran type inference only on the local expression it was currently considering. This PR uses the global information that works for more cases.

Blocked by ~#1259~, ~#1255~

### CODE REVIEW

File: src/gt4py/next/iterator/transforms/collapse_tuple.py [hunk 1]
```diff
@@ -80,7 +86,9 @@ def visit_FunCall(self, node: ir.FunCall, **kwargs) -> ir.Node:
                     # tuple argument differs, just continue with the rest of the tree
                     return self.generic_visit(node)
 
-            if self.ignore_tuple_size or _get_tuple_size(first_expr) == len(node.args):
+            if self.ignore_tuple_size or _get_tuple_size(self._node_types[id(first_expr)]) == len(
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/collapse_tuple.py:89]: I don't think that I want `self._node_types[id(first_expr)]` to become a pattern. Any thoughts?

File: src/gt4py/next/iterator/type_inference.py [hunk 2]
```diff
@@ -902,15 +896,34 @@ def visit_FencilDefinition(
...
+                    f"Expected a type to be inferred for node `{child_node}`, but none was found."
+                )
+
+
 def infer_all(
     node: ir.Node,
+    *,
     offset_provider: Optional[dict[str, Connectivity | Dimension]] = None,
     reindex: bool = True,
+    save_to_annex=False,
```
Comment (havogt) [inline src/gt4py/next/iterator/type_inference.py:918]: Why is this added? It's not used in this PR.

**Review by havogt — CHANGES_REQUESTED** [review]
Some questions

File: src/gt4py/next/iterator/type_inference.py (hunk 2 above)
Comment (havogt) [reply src/gt4py/next/iterator/type_inference.py:918]: Ok, it's in the other PR and reviewed separately.

File: src/gt4py/next/iterator/transforms/collapse_tuple.py [hunk 3]
```diff
@@ -56,8 +57,13 @@ def apply(
         If `ignore_tuple_size`, apply the transformation even if length of the inner tuple
         is greater than the length of the outer tuple.
         """
+        node_types = it_type_inference.infer_all(node)
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/collapse_tuple.py:60]: I don't know, but probably it is terrible :-(

File: src/gt4py/next/iterator/transforms/collapse_tuple.py (hunk 1 above)
Comment (havogt) [reply src/gt4py/next/iterator/transforms/collapse_tuple.py:89]: When you say annex in this context, you mean we guarantee that the type is always updated/preserved in all passes, right? Otherwise we would also have to run it before each pass.
Independent of the performance problem, is there any disadvantage in using a pass-local annex here?

Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/collapse_tuple.py:89]: With respect to the annex being "dangerous" yes. If it's only pass local you can still screw things up, but it is easier to find the problem (i.e. you do a transformation of the IR and some of the newly created nodes don't have a type or the type changed) as it is local to the pass.

Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/collapse_tuple.py:89]: See also this commit on how the code looks like if we use the annex: https://github.com/GridTools/gt4py/pull/1260/commits/a09f18f1d227c9a557a6930b7aa17fd897d08294

**Review by havogt — APPROVED** [review]
As discussed, we leave it as is for now.
